backends/k8s/client.go

```diff
@@ -0,0 +1,219 @@
+package k8s
+
+import (
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"fmt"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/pkg/runtime/schema"
+	"k8s.io/client-go/pkg/runtime/serializer"
+	"k8s.io/client-go/pkg/runtime"
+	"github.com/kelseyhightower/confd/log"
+	"github.com/projectcalico/libcalico-go/lib/backend/k8s/thirdparty"
+	kapiv1 "k8s.io/client-go/pkg/api/v1"
+	metav1 "k8s.io/client-go/pkg/apis/meta/v1"
+	clientapi "k8s.io/client-go/pkg/api"
+	kerrors "k8s.io/client-go/pkg/api/errors"
+	"time"
+	"regexp"
+	"encoding/json"
+	"github.com/projectcalico/libcalico-go/lib/net"
+)
+
+const (
+	ipPool          = "/calico/v1/ipam/v4/pool"
+	global          = "/calico/bgp/v1/global"
+	globalASN       = "/calico/bgp/v1/global/as_num"
+	globalNodeMesh  = "/calico/bgp/v1/global/node_mesh"
+	allNodes        = "/calico/bgp/v1/host"
+	globalLogging   = "/calico/bgp/v1/global/loglevel"
+)
+
+var (
+	singleNode = regexp.MustCompile("^/calico/bgp/v1/host/([a-zA-Z0-9._/-]*)")
+)
+
+type Client struct {
+	clientSet *kubernetes.Clientset
+	tprClient *rest.RESTClient
+}
+
+type KubeConfig struct {
+	Kubeconfig     string
+	K8sAPIEndpoint string
+	K8sKeyFile     string
+	K8sCertFile    string
+	K8sCAFile      string
+	K8sAPIToken    string
+}
+
+func NewK8sClient(kubeconfig string) (*Client, error) {
+
+	kc := KubeConfig {
+		Kubeconfig: kubeconfig,
+	}
+	log.Debug(fmt.Sprintf("Building client config for %+v", kc))
+
+	// Set an explicit path to the kubeconfig if one
+	// was provided.
+	loadingRules := clientcmd.ClientConfigLoadingRules{}
+	if kc.Kubeconfig != "" {
+		loadingRules.ExplicitPath = kc.Kubeconfig
+	}
+
+	// A kubeconfig file was provided.  Use it to load a config, passing through
+	// any overrides.
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&loadingRules, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the clientset
+	cs, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	log.Debug(fmt.Sprintf("Created k8s clientSet: %+v", cs))
+
+	tprClient, err := buildTPRClient(config)
+	if err != nil {
+		return nil, err
+	}
+	kubeClient := &Client{
+		clientSet: cs,
+		tprClient: tprClient,
+	}
+
+    	return kubeClient, nil
+}
+
+// Take the etcd like keys and route it to the appropriate k8s API endpoint
+func (c *Client) GetValues(keys []string) (map[string]string, error) {
+	var kvps = make(map[string]string)
+	for _, key := range keys {
+		log.Debug(fmt.Sprintf("Getting key %s", key))
+		if m := singleNode.FindStringSubmatch(key); m != nil {
+			host := m[len(m)-1]
+			node, err := c.clientSet.Nodes().Get(host, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+
+			var prefix = fmt.Sprintf("/calico/bgp/v1/host/%s/", node.Name)
+			for _, ip := range node.Status.Addresses {
+				if ip.Type == kapiv1.NodeInternalIP {
+					kvps[prefix + "ip_addr_v4"] = ip.Address
```

Comment:
Nope, let's skip all the ones I listed as optional up there, and not worry about IPv6 yet since k8s doesn't fully support it.

My main concern is about `/host/%s/network_v4`, which we'll need to populate to make the `cross-subnet` IPIP type work.

We've got a couple of options:
- Implement the `Apply()` action on the Node resource in Kubernetes and write this info as an annotation.
- Just hard code this to a /24 somehow (scary).

Let's chat about what this means.